File: Galerkin.m
clc
clear all


%%
%DEFINING THE PARAMETERS
dom =[0,1];

fprintf('Please enter the number of elements (default:10)\n')
Nelem = input('Number of elements: '); %NUMBER OF ELEMENTS
if isempty(Nelem)
    Nelem = 10; 
end

h = (dom(2) - dom(1))/Nelem;           %STEP SIZE
L = dom(2) - dom(1);                   %TOTAL LENGTH OF DOMAIN  
fprintf('Please enter the Peclet number (default:10)\n')
Pe = input('Peclet Number: ');         %PECLET NUMBER 
if isempty(Pe)
    Pe = 10; 
end

fprintf('Please enter the value of diffusion coefficient (default:0.01)\n')
nu = input('Diffusion Coefficient: '); %DIFFUSION COEFFICIENT
if isempty(nu)
    nu = 0.01; 
end

% Pe = a*h/(2*nu);
a = Pe*2*nu/h;                         %DEFINING 'a' IN TERMS OF Pe NUMBER
alpha = coth(Pe)-1/Pe;                 %DEFINING ALPHA 
nodes = Nelem + 1;                     %DEFINING THE NUMBER OF NODES 
X = (dom(1):h:dom(2))';                %DEFINING COORDINATES 
Conn = [1:(nodes-1);2:nodes]';         %CONNECTIVITY MATRIX
%alpha =1;

%%
%INITIALIZING THE MATRICES
K = zeros(nodes,nodes);                
F = zeros(nodes,1);

%DEFINING DIRICHLET NODES
nodes_Dir = [1,nodes];

%%
fprintf('Select a Method \n [1] Galerkin Method\n [2] Petrov Galerkin Method\n [3] Compare both\n')
choice = input('Method = ');

if choice==1                           %GALERKIN METHOD (SEPERATE)
    Method = 'Galerkin';
    for i = 1:Nelem                    %LOOP OVER ELEMENTS
        T = Conn(i,:);
        K_i = ((a/2)*[-1 1;-1 1] + (nu/h)*[1 -1;-1 1]);
        K(T,T) = K(T,T) + K_i;         %ASSEMBLING THE MATRIX
    end
else if choice==2                      %PETROV-GALERKIN METHOD (SEPERATE)
        Method = 'Petrov-Galerkin';
    for i = 1:Nelem                    %LOOP OVER ELEMENTS  
        T = Conn(i,:);   
        St = (alpha*a/2)*[1 -1;-1 1];  %ADDED STABILIZATION
        K_i = ((a/2)*[-1 1;-1 1] + (nu/h)*[1 -1;-1 1] + St);
        K(T,T) = K(T,T) + K_i;         %ASSEMBLING THE MATRIX
    end
else if choice==3                      %COMPARING BOTH METHODS TOGETHER
        Method='Comparison';
        K_G = zeros(nodes,nodes);
        F_G = zeros(nodes,1);
        K_P = zeros(nodes,nodes);
        F_P = zeros(nodes,1);
    for i = 1:Nelem                    %LOOP OVER ELEMENTS
        T = Conn(i,:);
        K_i_G = ((a/2)*[-1 1;-1 1] + (nu/h)*[1 -1;-1 1]);
        K_G(T,T) = K_G(T,T) + K_i_G;   %GALERKIN MATRIX ASSEMBLY
        K_i_P = ((a/2)*[-1 1;-1 1] + (nu/h)*[1 -1;-1 1] + (alpha*a/2)*[1 -1;-1 1]);
        K_P(T,T) = K_P(T,T) + K_i_P;   %PETROV-GALERKIN MATRIX ASSEMBLY
    end
else        
    error('Please select a valid method');
    end
    end
end

%%
%DETERMINING SOLUTION
if choice==1||choice==2
    F = F - K(:,nodes_Dir)*[1;0];       %TAKING DIRICHLET TERMS TO THE FORCE MATRIX SIDE
    K_f = K(2:nodes-1,2:nodes-1);       %FINAL K MATRIX 
    F_f = F(2:nodes-1);                 %FINAL F MATRIX

    sol = K_f\F_f;                      %SOLUTION
    u = zeros(nodes,1);                 %INITIALIZING THE SCALAR VARIABLE 
    u([1;nodes]) = [1;0];               %IMPOSING DIRICHLET VALUES 
    u(2:nodes-1) = sol;
    U_max = max(u);                     %FOR POST-PROCESSING    

else if choice==3
    
    %SOLUTION FOR GALERKIN METHOD 
    F_G = F_G - K_G(:,nodes_Dir)*[1;0]; 
    K_G_f = K_G(2:nodes-1,2:nodes-1); 
    F_G_f = F_G(2:nodes-1); 

    sol_G = K_G_f\F_G_f;
    u_G = zeros(nodes,1); 
    u_G([1;nodes]) = [1;0];
    u_G(2:nodes-1) = sol_G;         
    
    %SOLUTION FOR PETROV-GALERKIN METHOD
    F_P = F_P - K_P(:,nodes_Dir)*[1;0]; 
    K_P_f = K_P(2:nodes-1,2:nodes-1); 
    F_P_f = F_P(2:nodes-1); 

    sol_P = K_P_f\F_P_f;
    u_P = zeros(nodes,1); 
    u_P([1;nodes]) = [1;0];
    u_P(2:nodes-1) = sol_P;
    U_3max = max(u_G);
else
   error('Please select a valid method'); 
    end
end

%%
%POST-PROCESSING

%EXACT SOLUTION
U_exact =   (1 - (exp(a*(X-L)/nu)))/(1-exp(-(a*L/nu)));

%PLOTTING GALERKIN OR PETROV-GALERKIN DEPENDING ON USER CHOICE
if choice==1||choice==2 
    plot(X,u,'r-o',X,U_exact,'k:','LineWidth',1,'MarkerSize',5)
    l = legend(Method,'exact','Location','northwest'); 
    set(l,'FontSize',11)
    set(gca, 'FontSize',11);
    axis([0,1,0,(1.25*U_max)]);
    xlabel('Domain Length'); ylabel('Scalar Variable');
    t = sprintf('Pe = %s',num2str(Pe));
    title(t) 
    hold on
end    

%PLOTTING BOTH METHODS SIMULTANOUSLY FOR COMPARISON
if choice ==3
    plot(X,u_G,'r-o',X,u_P,'b-o',X,U_exact,'k:','LineWidth',1,'MarkerSize',5)
    l = legend('Galerkin','Petrov-Galerkin','exact','Location','northwest'); 
    set(l,'FontSize',11)
    set(gca, 'FontSize',11);
    axis([0,1,0,(1.25*U_3max)]);
    xlabel('Domain Length'); ylabel('Scalar Variable');
    t = sprintf('Pe = %s',num2str(Pe));
    title(t) 
end
    
%  if choice==1||choice==2 
%     plot(X,u,'b-o','LineWidth',1,'MarkerSize',5)
%     l = legend('exact','Upwind','Petrov-Galerkin','Location','northwest'); 
%     set(l,'FontSize',11)
%     set(gca, 'FontSize',11);
%     axis([0,1,0,(1.25*U_max)]);
%     xlabel('Domain Length'); ylabel('Scalar Variable');
%     t = sprintf('Pe = %s',num2str(Pe));
%     title(t) 
%     hold on
% end      

File: Galerkin.cpp
#include<iostream>
#include<cstdlib>
#include<cmath>
#include<iomanip>
#include<vector>
#include<fstream>
#include<cstring>

using namespace std;
using std::setw;
using std::ofstream;

double** inv(vector<vector<double>>  &arr);
double* mult(vector<vector<double>> &arr,vector<double> &array);

int main(){

	//DEFINING THE PARAMETERS
	double dom[2] = {0.0,1.0};
	int Nelem,nodes,i,j,k,p,q,T[2];	  
	double h,L,Pe,nu,a,alpha;
	typedef vector<vector<double>> matrixd;	//DEFINING A DOUBLE ARRAY MATRIX TYPE
	typedef vector<double> matrixs;   //DEFINING A SINGLE ARRAY MATRIX TYPE
	typedef vector<vector<int>> matrixint;   //DEFINING A DOUBLE ARRAY INT MATRIX TYPE
	double **K_invf;
	double *U_f;
	double K_i[2][2], St[2][2];
	int nodes_Dir[2];
	int choice, num = 21;
	double numer, denom;
	string str;
	ofstream output;
	system("cls");
	
	cout << "Please enter the number of elements" << endl;
	cout << "The number of elements: "; 
	cin >> Nelem; cout << endl;
	
	h = (dom[1] - dom[0])/Nelem;	//STEP SIZE
	L = dom[1] - dom[0];
	
	cout << "Please enter the Peclet number" << endl;
	cout << "Peclet number: "; 
	cin >> Pe; cout << endl;
	
	cout << "Please enter the value of diffusion coefficient" << endl;
	cout << "Diffusion coefficient: "; 
	cin >> nu; cout << endl;

	flag:
	cout << "Select the method: \n[1] Galerkin Method\n[2] Petrov Galerkin Method" << endl;
	cin >> choice; cout << endl;
	
	a = Pe*2.0*nu/h;		   //DEFINING 'a' IN TERMS OF Pe NUMBER	
	alpha = (1.0/tanh(Pe)) - (1.0/Pe);		//DEFINING ALPHA (OPTIMUM VALUE)
	nodes = Nelem + 1;

	//ALLOCATING DIRICHLET NODES
	nodes_Dir[0] = 0;
	nodes_Dir[1] = nodes-1;
	
	//DEFINING MATRICES
	matrixint conn(Nelem, vector<int>(2)); //DEFINING CONNECTIVITY MATRIX
	matrixs X(nodes);	//DEFINING COORDINATE MATRIX
	matrixd K(nodes, vector<double>(nodes)); //DEFINING STIFFNESS MATRIX
	matrixs F(nodes);	//DEFINING FORCE MATRIX
	matrixs U(nodes); 	//DEFINING SOLUTION MATRIX
	matrixs U_exact(nodes); //DEFINING EXACT SOLUTION
	matrixd K_f(nodes-2,vector<double>(nodes-2));	//DEFINING REDUCED K MATRIX
	matrixs F_f(nodes-2);	//DEFINING REDUCED F MATRIX
	matrixd K_inv(nodes-2, vector<double>(nodes-2));
	matrixd Result(nodes, vector<double>(3)); //RESULT FILE TO OUTPUT

	//BUILDING COORDINATE MATRIX
	for (i=0;i<nodes;i++){
		X[i] = i*h;
	}
	cout << endl;
    //BUILDING CONNECTIVITY MATRIX
	cout << "\nThe connectivity matrix: " << endl;
	cout << setw(5);
	for (i=0;i<Nelem;i++){
		for (j=0;j<2;j++){
			conn[i][j] = i + j +1;
			cout << conn[i][j] << setw(5);
		}
		cout << endl;
	}	

	switch (choice){
		case 1: //GALERKIN
		str = "Galerkin";
		K_i[0][0] = (-a/2) + (nu/h);
		K_i[0][1] = (a/2) + (-nu/h);
		K_i[1][0] = (-a/2) + (-nu/h);
		K_i[1][1] = (a/2) + (nu/h);
		break;

		case 2: //PETROV GALERKIN
		str = "Petrov-Galerkin";
		//STABILITY TERM
		St[0][0] = alpha*a/2;
		St[0][1] = -alpha*a/2;
		St[1][0] = -alpha*a/2;
		St[1][1] = alpha*a/2;

		K_i[0][0] = (-a/2) + (nu/h) + St[0][0];
		K_i[0][1] = (a/2) + (-nu/h) + St[1][0]; 
		K_i[1][0] = (-a/2) + (-nu/h) + St[1][0];
		K_i[1][1] = (a/2) + (nu/h) + St[1][1];
		break;

		default:
		cout << "Wrong input, please try again\n" << endl;
		goto flag;
	}
	
	//INITIALIZING THE K MATRIX
	for (i=0;i<nodes;i++){
		for (j=0;j<nodes;j++){
			K[i][j] = 0.0;
		}
	}	
	//INITIALIZING THE F MATRIX
	for (i=0;i<nodes;i++){
		F[i] = 0.0;
	}
	//INITIALIZING THE U MATRIX
	for (i=0;i<nodes;i++){
		U[i] = 0.0;
	}

	//ASSEMBLING THE GLOBAL STIFFNESS MATRIX MATRIX
	for (i=0;i<Nelem;i++){
		T[0] = (conn[i][0]) - 1;
		T[1] = (conn[i][1]) - 1;
		K[T[0]][T[0]] = K[T[0]][T[0]] + K_i[0][0];
		K[T[0]][T[1]] = K[T[0]][T[1]] + K_i[0][1];
		K[T[1]][T[0]] = K[T[1]][T[0]] + K_i[1][0];
		K[T[1]][T[1]] = K[T[1]][T[1]] + K_i[1][1];
	}
		
	//PRINTING THE GLOBAL STIFFNESS MATRIX
	if (Nelem < num){
		cout << "\n\nThe stiffness matrix is: " << endl;
		cout << setw(4);
		for (i=0;i<nodes;i++){
			for (j=0;j<nodes;j++){
				cout <<	K[i][j] << setw(4);
			}
			cout << endl;
		} 
	}
	else {
		cout << "The element count is too large to display matrix on screen..";
	}
	//----------------------------SOLUTION GENERATION------------------------------//

	//ASSIGNING DIRICHLET VALUES
	U[nodes_Dir[0]] = 1.0;
	U[nodes_Dir[1]] = 0.0;
	
	//BUILDING THE F MATRIX
	for (i=0;i<nodes;i++){
		F[i] = F[i] - ((K[i][nodes_Dir[0]] * U[nodes_Dir[0]]) +  (K[i][nodes_Dir[1]] * U[nodes_Dir[1]]));
	}
	//BUILDING REDUCED K MATRIX
	p=0;
	for(i=1;i<nodes-1;i++){
		q=0;
		for (j=1;j<nodes-1;j++){
			K_f[p][q] = K[i][j];
			q++;
		}
		p++;
	}
	//PRINTING REDUCED K MATRIX
	if (Nelem < num){
		cout << "\n\nThe reduced K matrix:"<< endl;
		cout << setw(4);
		for(i=0;i<nodes-2;i++){
			for (j=0;j<nodes-2;j++){
				cout << K_f[i][j] << setw(4);			
			}
			cout << endl;
		} 
	}

	//BUILDING REDUCED FORCE MATRIX
	p=0;
	for(i=1;i<nodes-1;i++){
		F_f[p] = F[i];
		p++;
	}

	//PRINT FORCE VECTOR
	if (Nelem < num){
		cout << "\n\nThe Force vector: " << endl;
		cout << setw(4);
		for(i=0;i<nodes;i++){
			cout << F[i] << setw(4)<< endl;
		}

	//PRINT REDUCED FORCE VECTOR
		cout << "\n\nThe reduced Force vector: " << endl;
		cout << setw(4);
		for(i=0;i<nodes-2;i++){
			cout << F_f[i] << setw(4)<< endl;
		} 
	}
	K_invf = inv(K_f); //INVERSE OBTAINED BY PERFORMING GAUSSIAN ELIMINATION

	//STRUCTURING THE INVERSE MATRIX
	for (i=0;i<nodes-2;i++){
        for (j=0;j<(nodes-2);j++){
    	    K_inv[i][j] = K_invf[i][j+(nodes-2)];
        }
    }
	cout << endl << endl << "Generating solution...\n\n";
	//PRINTING THE INVERSE STIFFNESS REDUCED MATRIX
	/*
		cout << "\n\nThe inverse of the reduced K matrix: " << endl;
		cout << setw(12);
		for (i=0;i<nodes-2;i++){
			for (j=0;j<nodes-2;j++){
				cout << K_inv[i][j] << setw(12);
			}
			cout << endl;
		}
	*/
	//OBTAINING SOLUTION FROM
	// [K_f][U_f] = [F_f] 	
	U_f = mult(K_inv,F_f);

	//ASSEMBLING THE SOLUTION MATRIX
	p = 0;
	for (i=1;i<nodes-1;i++){
		U[i] = U_f[p];
		p++;
	}
	//CALCULATING EXACT SOLUTION
	for (i =0;i<nodes;i++){
		numer = 1.0 - exp((a*(X[i]-L))/nu);
		denom = 1.0 - exp(-a*L/nu);
		U_exact[i] = numer/denom;
	}
	if (Nelem < 101){
		cout << str << setw(10) << "Exact" << endl;
			for (i=0;i<nodes;i++){
				cout <<setw(7)<< U[i] <<setw(15)<< U_exact[i] << endl;
			}
	}

	for (i=0;i<nodes;i++){
		Result[i][0] = U[i];
		Result[i][1] = U_exact[i];
		Result[i][2] = X[i];
	}

	//GENERATING OUTPUT FILE WITH SOLUTION AND NODAL COORDINATE
	output.open("Result.dat");
	if (!output){
		cerr << "Error opening file" << endl;
		exit(1);
	}
	output << str << "\tExact Solution\tCoordinate" << endl;
	for (i=0;i<nodes;i++){
		for (j=0;j<3;j++){
			output << Result[i][j] << "\t";
		}
		output << endl;
	}
	output.close();
	cout << "Results has been generated";

	delete[] K_invf;
	delete[] U_f;
	return 0;
}

double** inv(vector<vector<double>>  &arr){
	int i,j,k,n;
    int size1,size2,size;
    double sizemat,sizemat1,sizemat2;
    double **mat =NULL;
    double d = 0.0;
    n = arr.size();

    mat = new double*[2*n];
    for (i=0;i<2*n;++i){
        mat[i] = new double[2*n]();
    }
    
    cout << "\n\nGenerating inverse of the matrix.....";
    //Inputs the coefficients of the matrix
    for(i=0;i<n;i++){
        for(j=0;j<n;j++){
            mat[i][j] = arr[i][j];
        }
    }
    
    // Initializing Right-hand side to identity matrix
    for(i=0;i<n;i++){
        for(j=0;j<2*n;j++){
            if(j==(i+n)){
                mat[i][j] = 1;
            }
        }
    }
    
    // Partial pivoting
    for(i=n;i>1;i--){
        if(mat[i-1][1] < mat[i][1]){
            for(j=0;j<2*n;j++){
                d = mat[i][j];
                mat[i][j] = mat[i-1][j];
                mat[i-1][j] = d;
            }
        }
    }

    // Reducing To Diagonal Matrix
    for(i=0;i<n;i++){
        for(j=0;j<2*n;j++){
            if(j != i){
                d = mat[j][i] / mat[i][i];
                for(k=0;k<2*n;k++){
                    mat[j][k] -= mat[i][k]*d;
                }
            }
        }
    }
    
    // Reducing To Unit Matrix
    for(i=0;i<n;i++){
        d = mat[i][i];
        for(j=0;j<2*n;j++){
            mat[i][j] = mat[i][j]/d;
        }
    }
    return mat;
}	

double* mult(vector<vector<double>> &arrd,vector<double> &arrs){
	int i,j,n;
	double* mat = NULL;
	n = arrd.size();

	mat = new double[n];
	//PERFORMING MATRIX MULTIPLICATION
	//OF A 2-D MATRIX WITH A 1-D VECTOR
	for (i=0;i<n;i++){
			for (j=0;j<n;j++){
				mat[i] += arrd[i][j] * arrs[j]; 
			}
	}
	return mat;
}
